### PR TITLE
json-rpc: stronger typed RPCs

### PR DESCRIPTION
**Pull request by bmwill**

### CODE REVIEW

**Comment by libra-action** [comment]
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Chaning JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.

File: json-rpc/types/src/lib.rs [hunk 1]
```diff
@@ -2,12 +2,73 @@
...
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum Id {
+    /// Numeric id
+    Number(u64),
+    /// String id
+    String(Box<str>),
+}
+
```
Comment (CapCap) [inline json-rpc/types/src/lib.rs:29]:
```
impl std::fmt::Display for Id {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Id::Number(ref v) => std::fmt::Debug::fmt(v, f),
            Id::String(ref v) =>  std::fmt::Debug::fmt(v, f),
        }
    }
}

impl std::hash::Hash for Id {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        match self {
            Id::Number(ref v) => v.hash(state),
            Id::String(ref v) => v.hash(state),
        }
    }
}
```

**Review by xli — APPROVED** [review]
solid

File: json-rpc/types/src/lib.rs (hunk 1 above)
Comment (bmwill) [reply json-rpc/types/src/lib.rs:29]: So `Hash` is derived for Id. And yeah I suppose we could implement Display (although Debug is derived)

**Pull request closed without merging**